torch/csrc/lazy/core/lazy_graph_executor.cpp

```diff
@@ -872,12 +872,12 @@ void LazyGraphExecutor::BuildInputOutputAliases(
       auto it = output_tensor_id_map.find(data_info->tensor_id);
       if (it != output_tensor_id_map.end()) {
         size_t output_index = it->second;
-        const Shape& root_shape = lowering_ctx->GetResultShape(output_index);
-        if (Shape(parameters_data[i]->shape()) == root_shape &&
+        const c10::optional<Shape>& root_shape =
```

Comment:
I think for pytorch/xla we actually need to make `BuildInputOutputAliases` virtual instead of providing a `SetUpAlias` and `GetResultShape`. The reason being that we actually need to check `xla::shape` instead of `lazy::shape` to decide the aliasing.

For pytorch/xla, `lazy::shape` and `xla::shape` can diverge in some cases and that would cause an incorrect aliasing in this case.

Reply:
Got it, so what about replacing [this](https://github.com/pytorch/pytorch/blob/6e3d004f0c4ff01cc6b7c197fd9b8e1ffa1fe4e1/torch/csrc/lazy/core/lazy_graph_executor.cpp#L807) line with something like:
```
lowering_ctx->BuildInputOutputAliases(tensors, coll.indices);
```
Then the implementation of `BuildInputOutputAliases` can be defined in the lowering context.

Reply:
I think `BuildInputOutputAliases` should stay in the LazyGraphExecutor or we can open a new backend API for it. Lowering_context shouldn't take tensor directly, conceptually it only handle IR level and below.  `SetupAlias` is the right interface for the lowering_context.

Reply:
Thoughts on
```cpp
if (lowering_ctx->CheckResultShape(parameters_data[i], output_index) && alias_map[output_index] < 0){
    ...
}
```
It would reduce the duplicated code and also makes it flexible enough to be able to use `XLAShape` inside `CheckResultShape`

Reply:
Hey @JackCaoG what do you think of the idea Antonio proposed? With it we should be able to keep `SetUpAlias` as part of the `LoweringContext` interface as you suggested

Reply:
yea I am ok with this idea. I do have a question through. AFAIK buffer aliasing(at least the syntax here) is a very xla specified feature. Do you guys need it in the MLIR?

Reply:
MLIR maybe not. But our Cerebras lazy tensor backend which we've designed to subclass the MLIR backend does require buffer aliasing.